File: DFS&BFS/python/2644.py
import collections
n = int(input())
first_people , sec_people = map(int,input().split())
t = int(input())
graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)
for _ in range(t):
    a,b = map(int,input().split())
    graph[a].append(b)
    graph[b].append(a)
def bfs(graph,start,visited):
    q = collections.deque()
    q.append([start,0])
    visited[start] = True
    while q:
        idx,cnt = q.popleft()
        if idx == sec_people:
            return cnt
        for node in graph[idx]:
            if not visited[node]:
                q.append([node,cnt+1])
                visited[node] = True
    return -1
print(bfs(graph,first_people,visited))

File: DFS&BFS/javascript/2644.js
const input = require("fs").readFileSync("/dev/stdin").toString().split("\n");

const n = Number(input[0]);
const [start, end] = input[1].split(" ").map(Number);
const m = Number(input[2]);

const arr = Array.from(Array(n + 1), () => []);
let a, b;
for (let i = 3; i < m + 3; i++) {
  [a, b] = input[i].split(" ").map(Number);
  arr[a].push(b);
  arr[b].push(a);
}
const visited = new Array(n + 1).fill(false);
let result = Infinity;
const dfs = (node, depth) => {
  if (end === node) {
    result = Math.min(depth, result);
    return depth;
  }
  arr[node].forEach((nextNode) => {
    if (visited[nextNode]) {
      return;
    }
    visited[nextNode] = true;
    dfs(nextNode, depth + 1);
    visited[nextNode] = false;
  });
};
visited[start] = true;
dfs(start, 0);
console.log(result < Infinity ? result : -1);
